### DISCUSSION TITLE
Usage with React?

### DISCUSSION THREAD

**Discussion by devinhalladay** [antvis/X6#1794] [Q&A]
Hi, I'm really excited to use this project (and maybe contribute once I can read the docs!). Thanks for your hard work for the community.

I'm having a hard time finding any documentation or examples of how to use X6 with React. Specifically, I just can't seem to figure out how to actually render anything from X6 — where do I put the graph configuration code? I tried doing it in a `useEffect()` with no success. I also tried both `useRef()` and `document.getElementByID()` to supply a container, and X6 rejected both of those with an error that a container was not specified.

Anyways — Any chance you might be able to put together a demo of how to render a node graph within a React functional component? Thanks!

**Comment by NewByVector** [comment]
https://codesandbox.io/s/hooks-8d8cs ，does this help you?

**Reply by devinhalladay** [reply]
yes, thanks!

**Reply by devinhalladay** [reply]
is there any way to use `useRef()` instead of native JS element selectors? i am using next.js so even with `useEffect()` i get the error `ReferenceError: navigator is not defined`

**Reply by devinhalladay** [reply]
I was able to use next.js dynamic imports to solve this problem partially: `const GrovesCanvas = dynamic(() => import('../../components/Canvas'));`

But I still get the error on occasion when reloading the page. It seems, for some reason, x6 still tries to use `window.navigator` on the server, even when I combine both `useEffect( if (typeof window !== undefined && container.current && minimapContainer.current) { ... })` and dynamic import.

Is it possible that I need to use a cleanup function in my `useEffect`?

```
const container = useRef();
  const minimapContainer = useRef();
  const [graph, setGraph] = useState(null);

  useEffect(() => {
    if (
      typeof window !== undefined &&
      container.current &&
      minimapContainer.current
    ) {
      const g = new Graph({
        container: container.current,
        autoResize: true,
        grid: true,
        scroller: {
          enabled: true,
        },
        minimap: {
          enabled: true,
          container: minimapContainer.current,
          width: 208,
          height: 144,
        },
      });

      g.fromJSON(data);
      setGraph(g);
    }
  }, [container]);
```

**Reply by devinhalladay** [reply]
When the page hot-reloads on my local development server, the minimap is not re-rendered, it actually creates a new instance.

This screenshot shows what happens after two hot-reloads:
<img width="932" alt="image" src="https://user-images.githubusercontent.com/1934813/153767303-ee252372-ef24-4d28-a561-6519cb9fbcfc.png">

Seems to indicate that a useEffect cleanup is necessary?
